# Objective
Predict whether there will be Recession ( Bad ) or Not ( Good )

Yes = Recession , No = not recession

In [1]:
# importing required libraries 
import numpy as np 
import matplotlib.pyplot as plt 
import pandas as pd

In [2]:
# importing or loading the dataset 
Data = pd.read_csv(r"C:\Users\admin\Downloads\US-macroeconomic data 1996-2020\macrodata.csv")

Data

Unnamed: 0        RPI    INDPRO  CE16OV  UNRATE  PAYEMS  USGOOD  USTPU  \
0      1996-01   8909.327   74.6841  125125     5.6  118316   23196  23947   
1      1996-02   8983.863   75.8344  125639     5.5  118739   23280  23988   
2      1996-03   9015.588   75.7631  125862     5.5  118993   23276  24030   
3      1996-04   9039.466   76.4562  125994     5.6  119158   23316  24043   
4      1996-05   9078.928   77.0161  126244     5.6  119486   23358  24137   
..         ...        ...       ...     ...     ...     ...     ...    ...   
288    2020-02  17219.643  109.2966  158759     3.5  152463   21205  27830   
289    2020-03  16947.234  104.5221  155772     4.4  151090   21086  27723   
290    2020-04  19124.430   91.2658  133403    14.7  130303   18698  24475   
291    2020-05  18296.406   92.0613  137242    13.3  133028   19374  24858   
292    2020-06  17990.747   97.6032  142182    11.1  137809   19859  25852   

     HOUST  PERMIT  ...  EXUSUK  WPSFD49207  WPSID61  CPIAUCSL        SP500  \
0     1467    1387  ...  1.5288       129.7    125.5   154.700   636.020020   
1     1491    1420  ...  1.5360       129.7    125.0   155.000   640.429993   
2     1424    1437  ...  1.5271       130.5    125.3   155.500   645.500000   
3     1516    1463  ...  1.5160       130.9    125.7   156.100   654.169983   
4     1504    1457  ...  1.5152       130.9    126.2   156.400   669.119995   
..     ...     ...  ...     ...         ...      ...       ...          ...   
288   1567    1438  ...  1.2953       206.0    195.2   259.050  2954.219971   
289   1269    1356  ...  1.2369       202.4    192.2   257.953  2584.590088   
290    934    1066  ...  1.2420       196.3    185.3   255.902  2912.429932   
291   1038    1216  ...  1.2302       200.4    185.3   255.768  3044.310059   
292   1265    1258  ...  1.2523       201.1    187.7   257.214  3100.290039   

           NASDAQ    GOLDBAR     Regime      P/E  Dividend Yield  
0     1059.790039  20.754421     Normal  18.6845            2.26  
1     1100.050049  21.372637     Normal  18.8140            2.16  
2     1101.400024  21.460953     Normal  18.9630            2.18  
3     1190.520020  21.637589     Normal  18.7388            2.19  
4     1243.430054  22.255806     Normal  19.1670            2.15  
..            ...        ...        ...      ...             ...  
288   8567.370117  18.868309     Normal  25.3952            1.97  
289   7700.100098  18.215258  Recession  22.2177            2.25  
290   8889.549805  25.572948  Recession  25.0359            2.41  
291   9489.870117  23.862782  Recession  26.1696            2.10  
292  10058.769530  26.864592  Recession  26.6508            1.95  

[293 rows x 45 columns]

In [3]:
Data.columns

Index(['Unnamed: 0', 'RPI', 'INDPRO', 'CE16OV', 'UNRATE', 'PAYEMS', 'USGOOD',
       'USTPU', 'HOUST', 'PERMIT', 'DPCERA3M086SBEA', 'AMTMTI', 'AMTMNO',
       'ACOGNO', 'AMDMUO', 'BUSINV', 'ISRATIO', 'M1SL', 'M2SL', 'TOTRESNS',
       'BUSLOANS', 'REALLN', 'DTCTHFNM', 'FEDFUNDS', 'TB3MS', 'TB6MS', 'GS1',
       'GS5', 'GS10', 'AAA', 'TB3SMFFM', 'T1YFFM', 'AAAFFM', 'EXSZUS',
       'EXCAUS', 'EXUSUK', 'WPSFD49207', 'WPSID61', 'CPIAUCSL', 'SP500',
       'NASDAQ', 'GOLDBAR', 'Regime', 'P/E', 'Dividend Yield'],
      dtype='object')

In [4]:
Data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293 entries, 0 to 292
Data columns (total 45 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Unnamed: 0       293 non-null    object 
 1   RPI              293 non-null    float64
 2   INDPRO           293 non-null    float64
 3   CE16OV           293 non-null    int64  
 4   UNRATE           293 non-null    float64
 5   PAYEMS           293 non-null    int64  
 6   USGOOD           293 non-null    int64  
 7   USTPU            293 non-null    int64  
 8   HOUST            293 non-null    int64  
 9   PERMIT           293 non-null    int64  
 10  DPCERA3M086SBEA  293 non-null    float64
 11  AMTMTI           293 non-null    int64  
 12  AMTMNO           293 non-null    int64  
 13  ACOGNO           293 non-null    int64  
 14  AMDMUO           293 non-null    int64  
 15  BUSINV           293 non-null    int64  
 16  ISRATIO          293 non-null    float64
 17  M1SL            

# Taking specific column for Analysis

In [5]:
numeric_predictors=['PAYEMS', 'DPCERA3M086SBEA','M2SL', 'BUSLOANS','P/E' ]
X_train = Data[numeric_predictors]
y_train = Data['Regime']

# Columns Description-

- Commercial and Industrial Loans, All Commercial Banks - (BUSLOANS)
- M2SL M2 Money Stock
- Real personal consumption expenditures (chain-type quantity index)- (DPCERA3M086SBEA)
- PAYEMS All Employees: Total nonfarm
- S&P P/E ratio



In [6]:
X_train.head()

PAYEMS  DPCERA3M086SBEA    M2SL  BUSLOANS      P/E
0  118316           63.691  3647.9  721.8115  18.6845
1  118739           64.180  3661.8  724.1690  18.8140
2  118993           64.491  3687.0  727.3465  18.9630
3  119158           64.729  3697.8  732.0780  18.7388
4  119486           64.813  3709.6  736.4489  19.1670

In [7]:
Correlation='X_train.corr()'
Correlation.style.applymap(lambda x: 'background-color : yellow' if x > 0.5 else '')

AttributeError: 'str' object has no attribute 'style'

# PCA is effected by scale so you need to scale the features in the data before applying PCA.
You can transform the data onto unit scale (mean = 0 and variance = 1)

In [8]:
# performing preprocessing part 
from sklearn.preprocessing import StandardScaler 

sc = StandardScaler() # mean 0 & std=1
  
X_train_1 = sc.fit_transform(X_train)  # Fit to data, then transform it.

# Mean

In [9]:
X_train_1.mean(axis=0)

array([ 1.21253027e-15, -6.30515738e-16, -2.91007264e-16, -2.42506053e-16,
       -1.54597609e-16])

# Standard Deviation

In [10]:
X_train_1.std(axis=0)

array([1., 1., 1., 1., 1.])

In [11]:
X_train.head(3)

PAYEMS  DPCERA3M086SBEA    M2SL  BUSLOANS      P/E
0  118316           63.691  3647.9  721.8115  18.6845
1  118739           64.180  3661.8  724.1690  18.8140
2  118993           64.491  3687.0  727.3465  18.9630

In [12]:
X_train_1

array([[-2.17565967, -1.9577598 , -1.34484629, -1.30272928, -0.43354514],
       [-2.12053195, -1.92591948, -1.34094797, -1.29801005, -0.42558389],
       [-2.08742925, -1.9056693 , -1.33388053, -1.29164935, -0.41642383],
       ...,
       [-0.61344695,  0.40200486,  2.40544459,  3.12300945, -0.04308078],
       [-0.25830975,  0.95657332,  2.64341006,  3.3401358 ,  0.02661557],
       [ 0.36477685,  1.37531581,  2.7262843 ,  3.12402436,  0.05619826]])

# No. of variable in original dataset

In [13]:
# Applying PCA function on training 
# and testing set of X component 
from sklearn.decomposition import PCA 
  
model = PCA(n_components = 3) # we are specifying we want 3 PCA
  
principalComponents= model.fit(X_train_1) # Fit the model with X and apply the dimensionality reduction on X.

# Explained Variance

# The explained variance tells you how much information (variance) can be attributed to each of the principal components.

This is important as while you can convert high dimensional space to 2 dimensional space, you lose some of the variance (information) when you do this. By using the attribute explained_variance_ratio_,

# 

In [14]:
import numpy as np
list(np.round(principalComponents.explained_variance_,2))   # eigen values  >= 1

[3.76, 1.0, 0.17]

In [15]:
3.76/15

0.25066666666666665

In [16]:
list(np.round(principalComponents.explained_variance_ratio_,2))    # >= 50%

[0.75, 0.2, 0.03]

In [17]:
0.75 + 0.2 + 0.03

0.98

You can see that the first principal component contains 75% of the variance

The second principal component contains 20% of the variance & third PC contain 3%.

Together, the three components contain 98% of the information.

principalComponents.mean_

principalComponents.get_covariance()

In [18]:
X_train_1 ### scaled dataset

array([[-2.17565967, -1.9577598 , -1.34484629, -1.30272928, -0.43354514],
       [-2.12053195, -1.92591948, -1.34094797, -1.29801005, -0.42558389],
       [-2.08742925, -1.9056693 , -1.33388053, -1.29164935, -0.41642383],
       ...,
       [-0.61344695,  0.40200486,  2.40544459,  3.12300945, -0.04308078],
       [-0.25830975,  0.95657332,  2.64341006,  3.3401358 ,  0.02661557],
       [ 0.36477685,  1.37531581,  2.7262843 ,  3.12402436,  0.05619826]])

# Beta Values or Eigen Vector

Z1 = Beta1x1+Beta2X2 + beta3X3 + ..... + beta13X13

In [19]:
principalComponents.components_   ##  BETA VALUES  Z1 , Z2 & Z3 Z4

array([[ 0.48941354,  0.504066  ,  0.50533967,  0.497258  , -0.06130384],
       [ 0.00498751,  0.00304359,  0.00365569,  0.11081236,  0.99381745],
       [-0.73958039, -0.19236846,  0.46259475,  0.44697844, -0.04723974]])

z1=(0.48941354) + (0.504066) + (0.50533967)+(0.497258)+(-0.06130384) z1

# Renaming the Columns
Prediction for data to get data reduction

In [20]:
X_train_1  # scaled data

array([[-2.17565967, -1.9577598 , -1.34484629, -1.30272928, -0.43354514],
       [-2.12053195, -1.92591948, -1.34094797, -1.29801005, -0.42558389],
       [-2.08742925, -1.9056693 , -1.33388053, -1.29164935, -0.41642383],
       ...,
       [-0.61344695,  0.40200486,  2.40544459,  3.12300945, -0.04308078],
       [-0.25830975,  0.95657332,  2.64341006,  3.3401358 ,  0.02661557],
       [ 0.36477685,  1.37531581,  2.7262843 ,  3.12402436,  0.05619826]])

In [21]:
transformed_data=principalComponents.fit_transform(X_train_1)  ### only give data which is transformed ( principal component)
transformed_data

array([[-3.35245620e+00, -5.96949318e-01,  8.01756296e-01],
       [-3.30559773e+00, -5.88128222e-01,  7.58396488e-01],
       [-3.27301657e+00, -5.78067383e-01,  7.35698627e-01],
       [-3.24759553e+00, -5.90550241e-01,  7.23099595e-01],
       [-3.21950829e+00, -5.63177073e-01,  6.94630983e-01],
       [-3.19235032e+00, -5.59216726e-01,  6.72397152e-01],
       [-3.16107030e+00, -6.45817856e-01,  6.56282412e-01],
       [-3.13948782e+00, -6.24597217e-01,  6.38161637e-01],
       [-3.11499736e+00, -5.62076206e-01,  6.24749221e-01],
       [-3.07918419e+00, -6.14131643e-01,  6.09203068e-01],
       [-3.05122914e+00, -5.30900622e-01,  5.82670592e-01],
       [-3.02006609e+00, -5.54936358e-01,  5.71622719e-01],
       [-2.99160500e+00, -5.28769872e-01,  5.50506811e-01],
       [-2.95912453e+00, -5.19945119e-01,  5.25941849e-01],
       [-2.92148960e+00, -5.69424819e-01,  5.03498467e-01],
       [-2.89577544e+00, -5.09541533e-01,  4.81732780e-01],
       [-2.87512566e+00, -4.37423899e-01

In [22]:
print("original shape:   ", X_train_1.shape)
print("transformed shape:", transformed_data.shape)

original shape:    (293, 5)
transformed shape: (293, 3)


In [23]:
principalDf = pd.DataFrame(data = transformed_data
             , columns = ['Data PC1', 'Data PC2', 'Data PC3'])
principalDf.head()

Data PC1  Data PC2  Data PC3
0 -3.352456 -0.596949  0.801756
1 -3.305598 -0.588128  0.758396
2 -3.273017 -0.578067  0.735699
3 -3.247596 -0.590550  0.723100
4 -3.219508 -0.563177  0.694631

# To check the Principal Component is uncorrelated / independent

In [24]:
Correlation=round(principalDf.corr(),3)
Correlation.style.applymap(lambda x: 'background-color : yellow' if x > 0.5 else '')

# Identify variable is combining together in Which PCA

In [25]:
numeric_predictors_1=['PAYEMS','DPCERA3M086SBEA','M2SL','BUSLOANS','P/E']
scaled_train=pd.DataFrame(X_train_1,columns=numeric_predictors_1)
scaled_train.head()

PAYEMS  DPCERA3M086SBEA      M2SL  BUSLOANS       P/E
0 -2.175660        -1.957760 -1.344846 -1.302729 -0.433545
1 -2.120532        -1.925919 -1.340948 -1.298010 -0.425584
2 -2.087429        -1.905669 -1.333881 -1.291649 -0.416424
3 -2.065926        -1.890172 -1.330852 -1.282178 -0.430207
4 -2.023179        -1.884703 -1.327542 -1.273428 -0.403883

In [26]:
final_data=pd.concat([scaled_train,principalDf],axis=1)
final_data.head()

PAYEMS  DPCERA3M086SBEA      M2SL  BUSLOANS       P/E  Data PC1  \
0 -2.175660        -1.957760 -1.344846 -1.302729 -0.433545 -3.352456   
1 -2.120532        -1.925919 -1.340948 -1.298010 -0.425584 -3.305598   
2 -2.087429        -1.905669 -1.333881 -1.291649 -0.416424 -3.273017   
3 -2.065926        -1.890172 -1.330852 -1.282178 -0.430207 -3.247596   
4 -2.023179        -1.884703 -1.327542 -1.273428 -0.403883 -3.219508   

   Data PC2  Data PC3  
0 -0.596949  0.801756  
1 -0.588128  0.758396  
2 -0.578067  0.735699  
3 -0.590550  0.723100  
4 -0.563177  0.694631

In [27]:
Correlation_Matrix=pd.DataFrame(round(final_data.corr(),3))*100
Correlation_Matrix=abs(Correlation_Matrix.iloc[0:6,6:9])
Correlation_Matrix.style.applymap(lambda x: 'background-color : yellow' if x > 50 else '')

Data PC2 = P/E

# Submission to Business

# Coefficient Values

In [28]:
principalComponents.components_

array([[ 0.48941354,  0.504066  ,  0.50533967,  0.497258  , -0.06130384],
       [ 0.00498751,  0.00304359,  0.00365569,  0.11081236,  0.99381745],
       [-0.73958039, -0.19236846,  0.46259475,  0.44697844, -0.04723974]])

In [29]:
principalDf

Data PC1  Data PC2  Data PC3
0   -3.352456 -0.596949  0.801756
1   -3.305598 -0.588128  0.758396
2   -3.273017 -0.578067  0.735699
3   -3.247596 -0.590550  0.723100
4   -3.219508 -0.563177  0.694631
..        ...       ...       ...
288  3.997616  0.219663 -0.250313
289  3.959197  0.072338  0.256751
290  2.673557  0.310211  2.887059
291  3.350853  0.407866  2.721565
292  3.799477  0.418003  2.120532

[293 rows x 3 columns]

# Converting Unsupervised to Supervised -- Combining PC's & Y Variable

In [30]:
data=pd.concat([principalDf,y_train],axis=1)
data

Data PC1  Data PC2  Data PC3     Regime
0   -3.352456 -0.596949  0.801756     Normal
1   -3.305598 -0.588128  0.758396     Normal
2   -3.273017 -0.578067  0.735699     Normal
3   -3.247596 -0.590550  0.723100     Normal
4   -3.219508 -0.563177  0.694631     Normal
..        ...       ...       ...        ...
288  3.997616  0.219663 -0.250313     Normal
289  3.959197  0.072338  0.256751  Recession
290  2.673557  0.310211  2.887059  Recession
291  3.350853  0.407866  2.721565  Recession
292  3.799477  0.418003  2.120532  Recession

[293 rows x 4 columns]

# KNN

In [31]:
import numpy as np, pandas as pd
import matplotlib.pyplot as plt
from sklearn.neighbors import KNeighborsClassifier

# Step 1 : Data Partition

In [32]:
#Dividing data into train and test dataset
from sklearn.model_selection import train_test_split
#from random import seed

#seed(20)
train_x = data.drop(['Regime'],axis=1)
train_y = data['Regime']

# Train test split

X_train, X_test, y_train, y_test = train_test_split(train_x, train_y, test_size=0.3,random_state=20)

In [33]:
X_train.head()

Data PC1  Data PC2  Data PC3
205  0.732968 -0.447012  0.253586
119 -0.527331 -0.569629 -0.524837
95  -1.326900 -0.293789 -0.224735
222  1.437487 -0.333828  0.162082
132 -0.087569 -0.575155 -0.570189

# Scaling Data

In [34]:
from sklearn.preprocessing import StandardScaler ,minmax_scale
scaler = StandardScaler()   # converting data into Z-scale = X-mean / Standard Deviation
#scaler.fit(X_train)         # Mean = 0 & standard Deviation =1

X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [35]:
X_train[0:5]

array([[ 0.35608906, -0.43035354,  0.72352353],
       [-0.2903741 , -0.54421867, -1.39592795],
       [-0.70050855, -0.28806659, -0.57882571],
       [ 0.71746787, -0.32524797,  0.47438205],
       [-0.06480059, -0.5493498 , -1.51941067]])

In [36]:
X_train.mean(axis=0)

array([-4.33257766e-18, -2.59954659e-17,  2.16628883e-17])

In [37]:
X_train.std(axis=0)

array([1., 1., 1.])

# Model Building

In [38]:
# train the decision tree
## Instantiate the model with 5 neighbors. 
model = KNeighborsClassifier(n_neighbors=8)
## Fit the model on the training data.
model.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=8)

# Predictions on Train Dataset

In [39]:
X_train=pd.DataFrame(X_train,columns=['Data PC1','Data PC2','Data PC3'])
X_train.head()

Data PC1  Data PC2  Data PC3
0  0.356089 -0.430354  0.723524
1 -0.290374 -0.544219 -1.395928
2 -0.700509 -0.288067 -0.578826
3  0.717468 -0.325248  0.474382
4 -0.064801 -0.549350 -1.519411

In [40]:
y_train.reset_index(drop=True, inplace=True)

In [41]:
train=pd.concat([X_train,y_train],axis=1)
train.head()

Data PC1  Data PC2  Data PC3  Regime
0  0.356089 -0.430354  0.723524  Normal
1 -0.290374 -0.544219 -1.395928  Normal
2 -0.700509 -0.288067 -0.578826  Normal
3  0.717468 -0.325248  0.474382  Normal
4 -0.064801 -0.549350 -1.519411  Normal

In [42]:
train['Predicted']=model.predict(X_train)
train.head()

C:\Users\admin\anaconda3\lib\site-packages\sklearn\base.py:443: UserWarning: X has feature names, but KNeighborsClassifier was fitted without feature names
  warnings.warn(
C:\Users\admin\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)


Data PC1  Data PC2  Data PC3  Regime Predicted
0  0.356089 -0.430354  0.723524  Normal    Normal
1 -0.290374 -0.544219 -1.395928  Normal    Normal
2 -0.700509 -0.288067 -0.578826  Normal    Normal
3  0.717468 -0.325248  0.474382  Normal    Normal
4 -0.064801 -0.549350 -1.519411  Normal    Normal

# Model Performance Metrics

In [43]:
from sklearn.metrics import confusion_matrix
confusion_matrix = confusion_matrix(train['Regime'],train['Predicted'])
print(confusion_matrix)

[[184   2]
 [  8  11]]


In [44]:
Accuracy_Train=((184+11)/(205)*100)
print(Accuracy_Train)

95.1219512195122


# Accuracy of Train in KNN in Train is 95%

In [45]:
from sklearn.metrics import classification_report
print(classification_report(train['Regime'], train['Predicted']))

              precision    recall  f1-score   support

      Normal       0.96      0.99      0.97       186
   Recession       0.85      0.58      0.69        19

    accuracy                           0.95       205
   macro avg       0.90      0.78      0.83       205
weighted avg       0.95      0.95      0.95       205



# Step 8 : Predictions on Test Dataset

In [46]:
X_test=pd.DataFrame(X_test,columns=['Data PC1','Data PC2','Data PC3'])
X_test.head()

Data PC1  Data PC2  Data PC3
0  1.693160 -0.059338 -0.375587
1 -0.419190 -0.493405 -1.222543
2 -0.738305  0.385793 -0.753835
3 -0.231292 -0.537295 -1.504156
4  1.929037  0.372920  5.806758

In [47]:
y_test.reset_index(drop=True, inplace=True)

In [48]:
test=pd.concat([X_test,y_test],axis=1)
test.head()

Data PC1  Data PC2  Data PC3     Regime
0  1.693160 -0.059338 -0.375587     Normal
1 -0.419190 -0.493405 -1.222543     Normal
2 -0.738305  0.385793 -0.753835  Recession
3 -0.231292 -0.537295 -1.504156     Normal
4  1.929037  0.372920  5.806758  Recession

In [49]:
test['Predicted']=model.predict(X_test)
test.head()

C:\Users\admin\anaconda3\lib\site-packages\sklearn\base.py:443: UserWarning: X has feature names, but KNeighborsClassifier was fitted without feature names
  warnings.warn(
C:\Users\admin\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)


Data PC1  Data PC2  Data PC3     Regime Predicted
0  1.693160 -0.059338 -0.375587     Normal    Normal
1 -0.419190 -0.493405 -1.222543     Normal    Normal
2 -0.738305  0.385793 -0.753835  Recession    Normal
3 -0.231292 -0.537295 -1.504156     Normal    Normal
4  1.929037  0.372920  5.806758  Recession    Normal

# Step 9 : Model Performance Metrics on Test data

In [50]:
from sklearn.metrics import confusion_matrix
confusion_matrix = confusion_matrix(test['Regime'],test['Predicted'])
print(confusion_matrix)

[[76  1]
 [ 6  5]]


In [51]:
Accuracy_test=((76+5)/(88)*100)
Accuracy_test

92.04545454545455

# Accuracy of Train and Test data is 95% and 92% respectively.

In [52]:
from sklearn.metrics import classification_report
print(classification_report(test['Regime'], test['Predicted']))

              precision    recall  f1-score   support

      Normal       0.93      0.99      0.96        77
   Recession       0.83      0.45      0.59        11

    accuracy                           0.92        88
   macro avg       0.88      0.72      0.77        88
weighted avg       0.92      0.92      0.91        88



# Cross Validation Method - Random concept

To identify the best value of K

In [53]:
from sklearn.metrics import accuracy_score
for K in range(25):
    K_value = K+1
    model2 = KNeighborsClassifier(n_neighbors = K_value)
    model2.fit(X_train, y_train) 
    y_pred_Train= model2.predict(X_train)
    y_pred_test = model2.predict(X_test)
    print("Train Accuracy is ", 
          round(accuracy_score(y_train,y_pred_Train)*100,0), "      ","Test Accuracy is ", round(accuracy_score(y_test,y_pred_test)*100,0),"% for K-Value:",
          K_value)

Train Accuracy is  100.0        Test Accuracy is  97.0 % for K-Value: 1
Train Accuracy is  98.0        Test Accuracy is  95.0 % for K-Value: 2
Train Accuracy is  98.0        Test Accuracy is  94.0 % for K-Value: 3
Train Accuracy is  98.0        Test Accuracy is  94.0 % for K-Value: 4
Train Accuracy is  96.0        Test Accuracy is  92.0 % for K-Value: 5


C:\Users\admin\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\admin\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdim

Train Accuracy is  94.0        Test Accuracy is  92.0 % for K-Value: 6
Train Accuracy is  94.0        Test Accuracy is  91.0 % for K-Value: 7
Train Accuracy is  95.0        Test Accuracy is  92.0 % for K-Value: 8
Train Accuracy is  95.0        Test Accuracy is  91.0 % for K-Value: 9
Train Accuracy is  94.0        Test Accuracy is  91.0 % for K-Value: 10
Train Accuracy is  94.0        Test Accuracy is  91.0 % for K-Value: 11
Train Accuracy is  91.0        Test Accuracy is  89.0 % for K-Value: 12
Train Accuracy is  91.0        Test Accuracy is  89.0 % for K-Value: 13
Train Accuracy is  91.0        Test Accuracy is  88.0 % for K-Value: 14
Train Accuracy is  92.0        Test Accuracy is  89.0 % for K-Value: 15
Train Accuracy is  91.0        Test Accuracy is  88.0 % for K-Value: 16
Train Accuracy is  91.0        Test Accuracy is  88.0 % for K-Value: 17
Train Accuracy is  91.0        Test Accuracy is  88.0 % for K-Value: 18


C:\Users\admin\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\admin\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdim

Train Accuracy is  91.0        Test Accuracy is  88.0 % for K-Value: 19
Train Accuracy is  91.0        Test Accuracy is  88.0 % for K-Value: 20
Train Accuracy is  93.0        Test Accuracy is  89.0 % for K-Value: 21
Train Accuracy is  91.0        Test Accuracy is  88.0 % for K-Value: 22
Train Accuracy is  91.0        Test Accuracy is  89.0 % for K-Value: 23
Train Accuracy is  91.0        Test Accuracy is  88.0 % for K-Value: 24
Train Accuracy is  91.0        Test Accuracy is  88.0 % for K-Value: 25


C:\Users\admin\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)


# train the decision tree
Instantiate the model with 5 neighbors. model = KNeighborsClassifier(n_neighbors=8,metric='euclidean')

Fit the model on the training data. model.fit(X_train, y_train)

In [54]:
from sklearn.model_selection import GridSearchCV
param_grid = {'n_neighbors': np.arange(2, 9)}


knn2 = KNeighborsClassifier()
knn_gscv = GridSearchCV(knn2, 
                        param_grid, # 2,3,4,5,6,7,8
                        cv=10,
                        verbose=1)
knn_gscv.fit(X_train, y_train)

Fitting 10 folds for each of 7 candidates, totalling 70 fits


C:\Users\admin\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdims` to True or False to avoid this warning.
  mode, _ = stats.mode(_y[neigh_ind, k], axis=1)
C:\Users\admin\anaconda3\lib\site-packages\sklearn\neighbors\_classification.py:228: FutureWarning: Unlike other reduction functions (e.g. `skew`, `kurtosis`), the default behavior of `mode` typically preserves the axis it acts along. In SciPy 1.11.0, this behavior will change: the default value of `keepdims` will become False, the `axis` over which the statistic is taken will be eliminated, and the value None will no longer be accepted. Set `keepdim

GridSearchCV(cv=10, estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': array([2, 3, 4, 5, 6, 7, 8])},
             verbose=1)

In [55]:
#check top performing n_neighbors value
knn_gscv.best_params_

{'n_neighbors': 2}

In [56]:
knn_gscv.best_score_

0.961190476190476

# Logistic Regression

In [57]:
#Importing libraries
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O 
import os
import itertools
import warnings
warnings.filterwarnings("ignore")
import io

################# visualization #################
import matplotlib.pyplot as plt 
from PIL import  Image
import seaborn as sns 
import plotly.offline as py
py.init_notebook_mode(connected=True)
import plotly.graph_objs as go
import plotly.tools as tls
import plotly.figure_factory as ff

# Data Partition

In [58]:
#Dividing data into train and test dataset
from sklearn.model_selection import train_test_split
#from random import seed

#seed(20)
train_x = data.drop(['Regime'],axis=1)
train_y = data['Regime']

# Train test split

X_train, X_test, y_train, y_test = train_test_split(train_x, train_y, test_size=0.3,random_state=20)

# Model building

# Hypothesis
H0 :- There is no log linear relationship between Regime and all independent Variable
Vs

H1 :- There is log linear relationship between Regime and all independent Variable
Alpha = 0.05 (5%)

# Logistics Regression

In [59]:
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
logreg = LogisticRegression()
output=logreg.fit(X_train, y_train) # model

In [60]:
output.intercept_  # beto 

array([-2.59135299])

In [61]:
output.coef_  # Beta 1 -- 

array([[-0.06671285,  0.83760966, -0.38396724]])

In [62]:
output.feature_names_in_

array(['Data PC1', 'Data PC2', 'Data PC3'], dtype=object)

# Step 6 : Predictions on Train Dataset

In [63]:
train=pd.concat([X_train,y_train],axis=1)
train.head()

Data PC1  Data PC2  Data PC3  Regime
205  0.732968 -0.447012  0.253586  Normal
119 -0.527331 -0.569629 -0.524837  Normal
95  -1.326900 -0.293789 -0.224735  Normal
222  1.437487 -0.333828  0.162082  Normal
132 -0.087569 -0.575155 -0.570189  Normal

In [64]:
train['Predicted']=output.predict(X_train)
train.head()

Data PC1  Data PC2  Data PC3  Regime Predicted
205  0.732968 -0.447012  0.253586  Normal    Normal
119 -0.527331 -0.569629 -0.524837  Normal    Normal
95  -1.326900 -0.293789 -0.224735  Normal    Normal
222  1.437487 -0.333828  0.162082  Normal    Normal
132 -0.087569 -0.575155 -0.570189  Normal    Normal

# Step 7 : Model Performance Metrics

In [65]:
from sklearn.metrics import confusion_matrix
matrix = confusion_matrix(train['Predicted'], train['Regime'])
matrix

array([[183,  15],
       [  3,   4]], dtype=int64)

In [66]:
Accuracy_Train=((187)/(205)*100)
print(Accuracy_Train)

91.21951219512195


# Acccuracy of model is 91% so model is  good fit

In [67]:
from sklearn.metrics import classification_report
print(classification_report(train['Regime'],train['Predicted']))

              precision    recall  f1-score   support

      Normal       0.92      0.98      0.95       186
   Recession       0.57      0.21      0.31        19

    accuracy                           0.91       205
   macro avg       0.75      0.60      0.63       205
weighted avg       0.89      0.91      0.89       205



# Step 8 : Predictions on Test Dataset

In [68]:
test=pd.concat([X_test,y_test],axis=1)
test.head()

Data PC1  Data PC2  Data PC3     Regime
269  3.339627 -0.047479 -0.150091     Normal
112 -0.778461 -0.514910 -0.461157     Normal
63  -1.400585  0.431866 -0.289012  Recession
122 -0.412149 -0.562174 -0.564586     Normal
292  3.799477  0.418003  2.120532  Recession

In [69]:
test['Predicted']=output.predict(X_test)
test.head()

Data PC1  Data PC2  Data PC3     Regime Predicted
269  3.339627 -0.047479 -0.150091     Normal    Normal
112 -0.778461 -0.514910 -0.461157     Normal    Normal
63  -1.400585  0.431866 -0.289012  Recession    Normal
122 -0.412149 -0.562174 -0.564586     Normal    Normal
292  3.799477  0.418003  2.120532  Recession    Normal

# Step 9 : Model Performance Metrics on Test data

In [70]:
from sklearn.metrics import confusion_matrix
matrix = confusion_matrix(test['Predicted'],test['Regime'])
print(matrix)

[[76 10]
 [ 1  1]]


In [71]:
Accuracy_test=((77)/(88)*100)
Accuracy_test

87.5

In [72]:
from sklearn.metrics import classification_report
print(classification_report(test['Regime'],test['Predicted']))

              precision    recall  f1-score   support

      Normal       0.88      0.99      0.93        77
   Recession       0.50      0.09      0.15        11

    accuracy                           0.88        88
   macro avg       0.69      0.54      0.54        88
weighted avg       0.84      0.88      0.84        88



# Coefficient
For future Prediction below 4 variable data is mandatory

Note :- If anyone is not present output will not be reliable

In [73]:
Model_Values=pd.DataFrame(output.intercept_,index=['Intercept'],columns=['Coefficient'])
Model_Values

Coefficient
Intercept    -2.591353

In [74]:
abc=pd.DataFrame(output.coef_.T,index=output.feature_names_in_,columns=['Coefficient'])
abc

Coefficient
Data PC1    -0.066713
Data PC2     0.837610
Data PC3    -0.383967

In [75]:
Model_Values.append(abc)

Coefficient
Intercept    -2.591353
Data PC1     -0.066713
Data PC2      0.837610
Data PC3     -0.383967

# Finish